### ISSUE TITLE
reindex on full node restart

### ISSUE THREAD

**Issue by crazybits** [bitshares/bitshares-core#285]
always need to reindex the database when restart the full node

> node_1        | 2579956ms th_a       witness.cpp:114               plugin_initialize    ] witness plugin:  plugin_initialize() end
node_1        | 2579956ms th_a       application.cpp:396           startup              ] Detected unclean shutdown. Replaying blockchain...
node_1        | 2579956ms th_a       application.cpp:291           operator()           ] Initializing database...
node_1        | 2585453ms th_a       db_management.cpp:48          reindex              ] reindexing blockchain
node_1        | 2585454ms th_a       db_management.cpp:101         wipe                 ] Wiping database
node_1        | 2585471ms th_a       object_database.cpp:84        wipe                 ] Wiping object database...
node_1        | 2585471ms th_a       object_database.cpp:86        wipe                 ] Done wiping object databse.
node_1        | 2585471ms th_a       object_database.cpp:91        open                 ] Opening object database from /data_trusted_node/blockchain ...
node_1        | 2585472ms th_a       object_database.cpp:97        open                 ] Done opening object database.
node_1        | 2588939ms th_a       db_debug.cpp:82               debug_dump           ] total_balances[asset_id_type()].value: 12205723874026 core_asset_data.current_supply.value: 253409943611989 
node_1        | 2588946ms th_a       db_management.cpp:125         open                 ] last_block->id(): 0009aa1fcd3ecffc960c07ffc988e47236b2de55 last_block->block_num(): 633375 
node_1        | 2588946ms th_a       db_management.cpp:62          reindex              ] Replaying blocks...

**Comment by vikramrajkumar** [comment]
* What operating system?
* What is the output when you exit `witness_node`?
* What is inside `witness_node_data_dir/blockchain`?

If you are indeed shutting down cleanly, does the problem come back after restarting twice after deleting `witness_node_data_dir/blockchain/dblock`?

**Comment by oxarbitrage** [comment]
a single ctrl-c will exit the node clean keeping the blockchain safe. this can take a while until the blockchain in memory is dumped fully to disk. 
in low memory nodes i noted this process is sometimes impossible and you always need to replay. replay should be faster anyways than restarting with no blockchain at all.

hope it helps.

**Comment by crazybits** [comment]
@vikramrajkumar 

-  Ubuntu 15.04

- many block pushing errors prompted after sent ctrl+c

> 2607097ms asio       main.cpp:170                  operator()           ] Caught SIGINT attempting to exit cleanly
2607223ms th_a       main.cpp:183                  main                 ] Exiting from signal 2
2607246ms th_a       application.cpp:570           handle_block         ] Error when pushing block:
10 assert_exception: Assert Exception
item->num > std::max<int64_t>( 0, int64_t(_head->num) - (_max_size) ): attempting to push a block that is too old
    {"item->num":1,"head":195049,"max_size":20}
    th_a  fork_database.cpp:80 _push_block


> 2607355ms th_a       fork_database.cpp:66          push_block           ] Pushing block to fork database that failed to link: 0002ff266f0b0e190546fe72510015db98d10b63, 196390
2607355ms th_a       fork_database.cpp:67          push_block           ] Head: 195049, 0002f9e9fd1b75f104ba560174523452e9edbf5a
2607355ms th_a       application.cpp:567           handle_block         ] Error when pushing block:
3080000 unlinkable_block_exception: unlinkable block
block does not link to known chain
    {}
    th_a  fork_database.cpp:87 _push_block

> 2607722ms th_a       db_management.cpp:170         close                ] Database close unexpected exception: {"code":3070001,"name":"pop_empty_chain","message":"there are no blocks to pop","stack":[{"context":{"level":"error","file":"db_block.cpp","line":423,"method":"pop_block","hostname":"","thread_name":"th_a","timestamp":"2017-05-28T10:43:27"},"format":"there are no blocks to pop","data":{}},{"context":{"level":"warn","file":"db_block.cpp","line":431,"method":"pop_block","hostname":"","thread_name":"th_a","timestamp":"2017-05-28T10:43:27"},"format":"","data":{}}]}

- Yes, the problem persist even delete  `witness_node_data_dir/blockchain/dblock`

**Comment by abitmore** [comment]
Ubuntu 15.04 is not recommended. Best use 16.04 LTS, or 14.04 LTS.

**Comment by go-bitshares** [comment]
@abitmore @vikramrajkumar,typo error,i am using ubuntu 15.10

**Comment by crazybits** [comment]
@abitmore @vikramrajkumar sorry, type error, i am using ubuntu 15.10.

**Comment by vikramrajkumar** [comment]
What git revision are you using? Are you sure it is recent? Have you allowed the node to fully sync to the P2P network before exiting with Ctrl+C ?

**Comment by abitmore** [comment]
I got this error today after a fresh sync. Latest master, Ubuntu 16.04LTS, default shipped boost 1.58.

**Comment by abitmore** [comment]
When restarting with `--replay-blockchain` there is a message:
```
842014ms th_a       application.cpp:441           startup              ] Replaying blockchain due to: replay-blockchain argument specified
842014ms th_a       application.cpp:330           operator()           ] Initializing database...
847441ms th_a       db_management.cpp:51          reindex              ] reindexing blockchain
847442ms th_a       db_management.cpp:104         wipe                 ] Wiping database
847443ms th_a       db_management.cpp:170         close                ] Database close unexpected exception: {"code":10,"name":"assert_exception","message":"Assert Exception","stack":[{"context":{"level":"error","file":"index.hpp","line":111,"method":"get","hostname":"","thread_name":"th_a","timestamp":"2017-05-29T23:14:07"},"format":"maybe_found != nullptr: Unable to find Object","data":{"id":"2.1.0"}}]}
847447ms th_a       object_database.cpp:87        wipe                 ] Wiping object database...
```

**Comment by abitmore** [comment]
After finished the manual replay and caught up with the network, then  trying to quit, the "no blocks to pop" message appeared again. However, next time when start without `--replay-blockchain`, it won't replay automatically.

**Comment by crazybits** [comment]
@vikramrajkumar  i am using the latest revision of master branch(git commit : 19f0ec57f02523c1a63625326e38121e9f6203ff)

**Comment by abitmore** [comment]
@crazybits I got that same "no blocks to pop" message again with a new synced node. However, it doesn't automatically replay when restarting. Please make sure you press ctrl+c only once, and wait for the process to exit by itself, it may take a while if your RAM is not enough, please be patient, or try the latest release with proper options configured.

**Comment by abitmore** [comment]
According to the discussion in QQ group, the "need to replay" issue has gone, but the "no blocks to pop" message still appears.

**Comment by vikramrajkumar** [comment]
I haven't experienced the replay issue but I always get the "no blocks to pop message" when closing I believe. I think that is a separate issue.

**Comment by oxarbitrage** [comment]
lets close this issue and use https://github.com/bitshares/bitshares-core/issues/303 for the "no blocks to pop" message.
